### ISSUE TITLE
Libtorch with deeplabv3_resnet101 will not forward.

### ISSUE THREAD

**Issue by antithing** [pytorch/pytorch#23217]
## 🐛 Bug

When i convert the `deeplabv3_resnet101` model and use it in Libtorch, it gives a segmentation fault on forwarding.

## To Reproduce

1. convert the model using python:

```
import torch
import torchvision
from torchvision import models

deeplap_model = models.segmentation.deeplabv3_resnet101(pretrained=True)
deeplap_model.cuda()
deeplap_model.eval()

class wrapper(torch.nn.Module):
    def __init__(self, model):
        super(wrapper, self).__init__()
        self.model = model
    
    def forward(self, input):
        results = []
        output = self.model(input)
        for k, v in output.items():
            results.append(v)
        return tuple(results)

model = wrapper(deeplap_model)

example = torch.rand(1, 3, 224, 224).cuda()
#example = torch.rand(1, 3, 224, 224)
# Use torch.jit.trace to generate a torch.jit.ScriptModule via tracing.
traced_script_module = torch.jit.trace(model, example)


traced_script_module.save("model.pt")
```

2.

Load the model in c++ and run an image through it.

```
std::shared_ptr<torch::jit::script::Module> module = torch::jit::load("model.pt");
module->to(torch::kCUDA);
assert(module != nullptr);
std::cout << "ok\n";


std::vector<torch::jit::IValue> inputs;
cv::Mat image;
image = cv::imread("pic.jpeg", 1);
cv::Mat image_resized;
cv::resize(image, image_resized, cv::Size(224, 224));
cv::cvtColor(image_resized, image_resized, cv::COLOR_BGR2RGB);
cv::Mat image_resized_float;
image_resized.convertTo(image_resized_float, CV_32F, 1.0 / 255);

auto img_tensor = torch::from_blob(image_resized_float.data, { 1, 224, 224, 3 }, torch::kFloat32);
cout << "img tensor loaded..\n";
img_tensor = img_tensor.permute({ 0, 3, 1, 2 });
img_tensor[0][0] = img_tensor[0][0].sub(0.485).div(0.229);
img_tensor[0][1] = img_tensor[0][1].sub(0.456).div(0.224);
img_tensor[0][2] = img_tensor[0][2].sub(0.406).div(0.225);

// to GPU
img_tensor = img_tensor.to(at::kCUDA);

torch::Tensor out_tensor2 = module->forward({ img_tensor }).toTensor(); //SEGFAULT
```

3.

Try with dummy test data:

```
std::shared_ptr<torch::jit::script::Module> module = torch::jit::load("model.pt");
	module->to(torch::kCUDA);
	assert(module != nullptr);
	std::cout << "ok\n";


	std::vector<torch::jit::IValue> inputs;
	inputs.emplace_back(torch::rand({ 64, 3, 224, 224 }));

	module->forward(inputs).toTensor();  //SEGFAULT
```

## Expected behavior

I would expect that i could run images, or data through the converted model in c++ and see the segmentation results.

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Microsoft Windows 10 Home
GCC version: Could not collect
CMake version: version 3.12.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.0\bin\cudnn64_7.dll

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.1.0
[pip3] torchvision==0.3.0

**Comment by peterjc123** [comment]
It seems that the tensor in the code block #3 is not passed to GPU.

**Comment by antithing** [comment]
Updated to this:

st```
d::shared_ptr<torch::jit::script::Module> module = torch::jit::load("model.pt");
	module->to(torch::kCUDA);
	assert(module != nullptr);
	std::cout << "ok\n";

	torch::Tensor img_tensor = torch::rand({ 64, 3, 224, 224 });
	img_tensor = img_tensor.to(at::kCUDA);

	torch::Tensor out_tensor2 = module->forward({ img_tensor }).toTensor();

```

I see the same fault. Is this correct? Thank you.

**Comment by peterjc123** [comment]
Would you please try this one: https://github.com/pytorch/pytorch/issues/19302#issuecomment-484769457?

**Comment by antithing** [comment]
Thank you! that compiles and runs.

How should i do the same thing to the real data?

```
	cv::Mat image;
	image = cv::imread("pic.jpeg", 1);
	cv::Mat image_resized;
	cv::resize(image, image_resized, cv::Size(224, 224));
	cv::cvtColor(image_resized, image_resized, cv::COLOR_BGR2RGB);
	cv::Mat image_resized_float;
	image_resized.convertTo(image_resized_float, CV_32F, 1.0 / 255);
	torch::Tensor img_tensor = torch::from_blob(image_resized_float.data, { 1, image_resized_float.rows, image_resized_float.cols, 3 }, torch::kF32);
```
Thanks again.

**Comment by peterjc123** [comment]
What about changing `img_tensor = img_tensor.to(at::kCUDA);` to `img_tensor.to(at::kCUDA);`?

**Comment by antithing** [comment]
Adjusted to this:

```
        auto module = torch::jit::load("model.pt", torch::kCUDA);
	assert(module != nullptr);
	std::cout << "ok\n";

	cv::Mat image;
	image = cv::imread("pic.jpeg", 1);
	cv::Mat image_resized;
	cv::resize(image, image_resized, cv::Size(224, 224));
	cv::cvtColor(image_resized, image_resized, cv::COLOR_BGR2RGB);
	cv::Mat image_resized_float;
	image_resized.convertTo(image_resized_float, CV_32F, 1.0 / 255);
	auto img_tensor = torch::from_blob(image_resized_float.data, { 1, 3, 224, 224 });

	cout << "img tensor loaded..\n";
	img_tensor = img_tensor.permute({ 0, 3, 1, 2 });
	img_tensor[0][0] = img_tensor[0][0].sub(0.485).div(0.229);
	img_tensor[0][1] = img_tensor[0][1].sub(0.456).div(0.224);
	img_tensor[0][2] = img_tensor[0][2].sub(0.406).div(0.225);

	// to GPU
	img_tensor.to(at::kCUDA);


	std::cout << "forwarding... " << std::endl;
	torch::Tensor out_tensor2 = module->forward({ img_tensor }).toTensor();
	std::cout << "forwarded! " << std::endl;
```  

I still get the crash at `forward`. Where am i going wrong?

**Comment by peterjc123** [comment]
Then would you please try this?
```
auto img_tensor = torch::from_blob(image_resized_float.data, { 1, 3, 224, 224 });
// to
auto img_tensor = torch::from_blob(image_resized_float.data, { 1, 3, 224, 224 }, torch::kCUDA);
// and then remove the following line
img_tensor.to(at::kCUDA);
```
The problem is clearly written in the link I sent you. There is a bug when you try to do forward with a tensor that you switched back/forth CPU and GPU.

**Comment by antithing** [comment]
This line:

`auto img_tensor = torch::from_blob(image_resized_float.data, { 1, 3, 224, 224 }, torch::kCUDA);`

crashes with an exception.
`
Unhandled exception at 0x00007FF8F4E1A388 in Segmentation.exe: Microsoft C++ exception: c10::Error at memory location 0x000000D98752E7A0. occurred`

at `C:\ThirdParty\Torch\release\libtorch\include\ATen\Functions.h`

Line 1542
` if (options.device().has_index()) {`

**Comment by peterjc123** [comment]
Sorry for keep you retrying. What's the result if you change the code to the following?
```c++
        auto module = torch::jit::load("model.pt", torch::kCUDA);
	assert(module != nullptr);
	std::cout << "ok\n";

	cv::Mat image;
	image = cv::imread("pic.jpeg", 1);
	cv::Mat image_resized;
	cv::resize(image, image_resized, cv::Size(224, 224));
	cv::cvtColor(image_resized, image_resized, cv::COLOR_BGR2RGB);
	cv::Mat image_resized_float;
	image_resized.convertTo(image_resized_float, CV_32F, 1.0 / 255);
	auto img_tensor = torch::from_blob(image_resized_float.data, { 1, 3, 224, 224 });

	cout << "img tensor loaded..\n";
	img_tensor = img_tensor.permute({ 0, 3, 1, 2 });
	img_tensor[0][0] = img_tensor[0][0].sub(0.485).div(0.229);
	img_tensor[0][1] = img_tensor[0][1].sub(0.456).div(0.224);
	img_tensor[0][2] = img_tensor[0][2].sub(0.406).div(0.225);

	// to GPU
	torch::Tensor img_tensor_gpu = img_tensor.to(at::kCUDA);


	std::cout << "forwarding... " << std::endl;
	torch::Tensor out_tensor2 = module->forward({ img_tensor_gpu }).toTensor();
	std::cout << "forwarded! " << std::endl;
```

**Comment by antithing** [comment]
Ok, we are back to a crash at this line:

`torch::Tensor out_tensor2 = module->forward({ img_tensor_gpu }).toTensor();`

Thanks again for your time and patience!

**Comment by peterjc123** [comment]
I'm out of ideas. @yf225 Do you have any ideas?

**Comment by antithing** [comment]
Still stuck on this! Did anyone have any more thoughts as to what I can try? Thank you!

**Comment by Jelly123456** [comment]
@antithing did you find the solution to this problem regarding processing real-data? I am facing the same problem now.

**Comment by nbiresev** [comment]
I got the same issue. The problem is that your model likely gets as output a tuple, which contains segmentation mask as 1st element of tuple. You can't convert tuple to Tensor that is why it crashes. 

```
c10::intrusive_ptr<c10::ivalue::Tuple> outputTuple = model.forward(inputs).toTuple();
at::Tensor outputTensor = outputTuple->elements()[0].toTensor();
```

This works for me.

**Comment by taras-janea** [comment]
The issue is that `module->forward(inputs)` returns a Tuple, not a single Tensor.
Below is a code snippet with the fix applied:

```cpp
torch::jit::script::Module module = torch::jit::load("model.pt");
module.to(torch::kCUDA);
module.eval();

cv::Mat image = cv::imread("pic.jpeg", 1);
cv::Mat image_resized;
cv::resize(image, image_resized, cv::Size(224, 224));
cv::cvtColor(image_resized, image_resized, cv::COLOR_BGR2RGB);
cv::Mat image_float;
image_resized.convertTo(image_float, CV_32F, 1.0 / 255);

auto img_tensor = torch::from_blob(
    image_float.data, {1, 224, 224, 3}, torch::kFloat32
).clone();

img_tensor = img_tensor.permute({0, 3, 1, 2});
img_tensor[0][0] = img_tensor[0][0].sub(0.485).div(0.229);
img_tensor[0][1] = img_tensor[0][1].sub(0.456).div(0.224);
img_tensor[0][2] = img_tensor[0][2].sub(0.406).div(0.225);
img_tensor = img_tensor.to(torch::kCUDA);

std::vector<torch::jit::IValue> inputs;
inputs.push_back(img_tensor);

// Convert to tuple & access output tensor
auto output_tuple = module.forward(inputs).toTuple();
torch::Tensor outputTensor = output_tuple->elements()[0].toTensor();
```

@antithing, if the issue is still reproducible on your setup using the provided code snippet with the latest PyTorch version, then please share:
- output of [script](https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py);
- complete, self-contained reproducer program (compilable without external dependencies) that demonstrates the issue, along with its output when run using the latest PyTorch version.

**Comment by taras-janea** [comment]
Closing this issue as it is no longer reproducible with the latest PyTorch version. If you continue to experience this problem or have related concerns, please feel free to open a new issue.
